Fix Redis metric types; add server_name label

pmcollins opened the pull request:
In preparing for a demo I noticed that some metric types were wrong (might have fallen through the cracks when I was dealing with merge issues on the Redis PR). Fixed those and added an optional server_name config value that will get applied as a Resource label, if present, for distinguishing between multiple Redis servers.

Commit:
Fix Redis metric types; add server_name label

bogdandrutu commented on receiver/redisreceiver/proto.go:
```diff
@@ -24,11 +24,15 @@ import (
+func newMetricsData(protoMetrics []*metricspb.Metric, serverName string) *consumerdata.MetricsData {
+	labels := map[string]string{"type": typeStr}
+	if serverName != "" {
+		labels["server_name"] = serverName
```
We should follow the semantic conventions defined here (if this can be used as service name):
https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions#service

pmcollins replied:
I wonder if it makes sense to define these names somewhere so that developers can browse and use them without the possibility of making a mistake.

bogdandrutu replied:
They are defined here: https://docs.google.com/presentation/d/1Q6CjinzIG5D6BweCpj-_AFQG0V0NmVkAgVPfhj0G8XQ/edit?usp=sharing

pmcollins replied:
Thank you for the link Bogdan, I will take a look. What I was wondering was if it might be helpful to capture these conventions in a package so that developers can just say, for example, `conventions.ServiceName` or something rather than `"service.name"`.

bogdandrutu commented on receiver/redisreceiver/proto.go:
```diff
@@ -24,11 +24,15 @@ import (
 
-func newMetricsData(protoMetrics []*metricspb.Metric) *consumerdata.MetricsData {
+func newMetricsData(protoMetrics []*metricspb.Metric, serverName string) *consumerdata.MetricsData {
+	labels := map[string]string{"type": typeStr}
```
No need to add this.

bogdandrutu commented on receiver/redisreceiver/proto.go:
```diff
@@ -46,7 +50,7 @@ func buildKeyspaceKeysMetric(k *keyspace, t *timeBundle) *metricspb.Metric {
 		name:   "redis/db/keys",
 		labels: map[string]string{"db": k.db},
-		mdType: metricspb.MetricDescriptor_CUMULATIVE_INT64,
+		mdType: metricspb.MetricDescriptor_GAUGE_INT64,
```
@pmcollins can you add description for each metric? It is hard to know when one is gauge or cumulative.

bogdandrutu commented on receiver/redisreceiver/redis_runnable.go:
```diff
@@ -104,7 +107,7 @@ func (r *redisRunnable) Run() error {
+	md := newMetricsData(metrics, r.serverName)
 
 	err = r.metricsConsumer.ConsumeMetricsData(r.ctx, *md)
 	numTimeSeries, numPoints := obsreport.CountMetricPoints(*md)
```
You can simply defer obsreport.EndMetricsReceiveOp(ctx, dataformat, numPoints, numTimeSeries, err) instead of adding it every time you return.

pmcollins replied:
I punted on this one for now. Hoping to discuss with you tomorrow.

Commit:
Address PR feedback: comments, serviceName

bogdandrutu commented on receiver/redisreceiver/config.go:
```diff
@@ -24,6 +24,9 @@ type config struct {
 	CollectionInterval time.Duration `mapstructure:"collection_interval"`
+	// The logical name of the Redis server. This value will be added as a
+	// "service.name" Resource label.
+	ServiceName string `mapstructure:"service_name"`
```
Should this be "service.name" for consistency? Also maybe leave a TODO to allow users to configure more things in the `Resource`. Something like:

```
   Resource map[string]AttributeValue 
```

Also can be done in a separate PR :)

pmcollins replied:
Agreed. Done.

pmcollins commented on receiver/redisreceiver/metric_functions.go:
```diff
@@ -72,45 +70,50 @@ func getDefaultRedisMetrics() []*redisMetric {
 	}
 }
 
+// User CPU consumed by the background processes in seconds since server start
```
Oh! Fixed. :-)

Commit:
Address PR feedback for metric descriptions

Moved metric descriptions from comments to metadata.
Changed service_name config field to service.name.

Commit:
Fix formatting

bogdandrutu commented on receiver/redisreceiver/proto.go:
```diff
@@ -24,11 +24,11 @@ import (
 		Resource: &resourcepb.Resource{
 			Type:   typeStr,
-			Labels: map[string]string{"type": typeStr},
+			Labels: map[string]string{"service.name": serviceName},
```
Probably you should add this only if not empty?

pmcollins replied:
Yes I can do that. I was thinking this would be a required field (I put that in the docs, too, but can take it out).

bogdandrutu replied:
Ok, we need to revisit this anyway, as I mentioned I think we should let user define more attributes in the `Resource` not just this one.

bogdandrutu commented:
Please rebase, the build should no longer fail :)

pmcollins committed:
Replace dot in service.name with underscore
